perfkitbenchmarker/configs/benchmark_config_spec.py

```diff
@@ -0,0 +1,335 @@
+        value, component_full_name, flag_values)
+    return static_virtual_machine.StaticVmSpec(
+        self._GetOptionFullName(component_full_name), **input_dict)
+
+
+class _StaticVmListDecoder(option_decoders.ListDecoder):
+  """Decodes the static_vms list of a VM group config object."""
+
+  def __init__(self, **kwargs):
+    super(_StaticVmListDecoder, self).__init__(
+        default=list, item_decoder=_StaticVmDecoder(), **kwargs)
+
+
+class _VmGroupSpec(spec.BaseSpec):
+  """Configurable options of a VM group.
+
+  Attributes:
+    cloud: string. Cloud provider of the VMs in this group.
+    disk_count: int. Number of data disks to attach to each VM in this group.
+    disk_spec: BaseDiskSpec. Configuration for all data disks to be attached to
+        VMs in this group.
+    os_type: string. OS type of the VMs in this group.
+    static_vms: None or list of StaticVmSpecs. Configuration for all static VMs
+        in this group.
+    vm_count: int. Number of VMs in this group, including static VMs and
+        provisioned VMs.
+    vm_spec: BaseVmSpec. Configuration for provisioned VMs in this group.
+  """
+
+  def __init__(self, component_full_name, flag_values=None, **kwargs):
+    super(_VmGroupSpec, self).__init__(component_full_name,
+                                       flag_values=flag_values, **kwargs)
+    providers.LoadProvider(self.cloud.lower())
+    if self.disk_spec:
+      disk_config = getattr(self.disk_spec, self.cloud, None)
+      if disk_config is None:
+        raise errors.Config.MissingOption(
+            '{0}.cloud is "{1}", but {0}.disk_spec does not contain a '
+            'configuration for "{1}".'.format(component_full_name, self.cloud))
+      disk_spec_class = disk.GetDiskSpecClass(self.cloud)
+      self.disk_spec = disk_spec_class(
+          '{0}.disk_spec.{1}'.format(component_full_name, self.cloud),
+          flag_values=flag_values, **disk_config)
+    vm_config = getattr(self.vm_spec, self.cloud, None)
+    if vm_config is None:
+      raise errors.Config.MissingOption(
+          '{0}.cloud is "{1}", but {0}.vm_spec does not contain a '
+          'configuration for "{1}".'.format(component_full_name, self.cloud))
+    vm_spec_class = virtual_machine.GetVmSpecClass(self.cloud)
+    self.vm_spec = vm_spec_class(
+        '{0}.vm_spec.{1}'.format(component_full_name, self.cloud),
+        flag_values=flag_values, **vm_config)
+
+  @classmethod
+  def _GetOptionDecoderConstructions(cls):
+    """Gets decoder classes and constructor args for each configurable option.
+
+    Returns:
+      dict. Maps option name string to a (ConfigOptionDecoder class, dict) pair.
+      The pair specifies a decoder class and its __init__() keyword arguments
+      to construct in order to decode the named option.
+    """
+    result = super(_VmGroupSpec, cls)._GetOptionDecoderConstructions()
+    result.update({
+        'cloud': (option_decoders.EnumDecoder, {
```

Comment:
I'm omitting this for now because it would be non-uniform with regard to the other items in the dict, but it's a good idea for a future change. The string 'cloud' is quite common in PKB code, and there may be a better way to avoid duplication.